### PR TITLE
Default to parallel image pulls with docker 1.9+.

### PR DESCRIPTION
**Pull request by aveshagarwal**

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@ncdc

**Comment by ncdc** [comment]
@aveshagarwal I'll try to take a look at this tomorrow if there's any down time from the all day meetings. In the meantime, if either @DirectXMan12 or @ingvagabund have some time, perhaps they can review.

File: pkg/cmd/server/kubernetes/node_config.go [hunk 1]
```diff
@@ -280,3 +288,30 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
...
+
+	env, err := dockerClient.Version()
+	if err != nil {
+		return false
+	}
+
+	serverVersionString := env.Get("ApiVersion")
+	serverVersion, err := dockerclient.NewAPIVersion(serverVersionString)
+	if err != nil {
+		return false
```
Comment (ingvagabund) [inline pkg/cmd/server/kubernetes/node_config.go]: glog.Warningf("Unable to determine Docker server version from %q.\n%v", serverVersionString, err)

File: pkg/cmd/server/kubernetes/node_config.go [hunk 2]
```diff
@@ -280,3 +288,30 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
...
+	}
+
+	minimumVersion, err := dockerclient.NewAPIVersion(version)
+	if err != nil {
+		return false
```
Comment (ingvagabund) [inline pkg/cmd/server/kubernetes/node_config.go]: glog.Warningf("Unable to determine minimal Docker server version from %q.\n%v", serverVersionString, err)

File: pkg/cmd/server/kubernetes/node_config.go (hunk 1 above)
Comment (DirectXMan12) [inline pkg/cmd/server/kubernetes/node_config.go]: :+1:

File: pkg/cmd/server/kubernetes/node_config.go (hunk 2 above)
Comment (DirectXMan12) [inline pkg/cmd/server/kubernetes/node_config.go]: :+1:

**Comment by DirectXMan12** [comment]
LGTM besides the lack of warnings.

File: pkg/cmd/server/kubernetes/node_config.go (hunk 1 above)
Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Don't use \n, use :

**Comment by ingvagabund** [comment]
ensureDockerAPIVersionForParallelImagePulls is configuration related function so node_config.go
 seems to be the best to put it in. I don't think it is general enough to be exported and put somewhere to github.com/openshift/origin/pkg/cmd/util/docker. Otherwise there is another scenario asking for minimal version for pulls. However, it can be retrieved from node configuration later.

File: pkg/cmd/server/kubernetes/node_config.go [hunk 3]
```diff
@@ -280,3 +288,30 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 
 	return config, nil
 }
+
+// ensureDockerAPIVersionForParallelImagePulls returns true if the docker API version
+// is >=1.21, else false.
+func ensureDockerAPIVersionForParallelImagePulls(dh *docker.Helper, version string) bool {
```
Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Why do you have this taking in a `docker.Helper`? Why not just take the client? Or maybe just take an interface that just contains the `Version` func?

**Comment by ncdc** [comment]
cc @liggitt @deads2k fyi... I plan to review this later, but wanted to make you aware in case you feel like getting to it first :smile:

File: pkg/cmd/server/kubernetes/node_config.go (hunk 3 above)
Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: At this point, I dont have access to docker client, so I have to create it using docker helper.

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: You can create the helper here, and pass in the client

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: You can create the helper in `BuildKubernetesNodeConfig`, get the client, and pass the client here.

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: I am trying to understand, BuildKubernetesNodeConfig calls this ensureDockerAPIVersionForParallelImagePulls, which takes helper and create the client. If BuildKubernetesNodeConfig creates the client by the helper, and the pass the client to  ensureDockerAPIVersionForParallelImagePulls, what would be the net benefit?

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: One advantage I see now would be that it would be easier to reuse/export that function later if needed. Is that what you are thinking too?

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: It seems to me that it would be better to add this kind of function to github.com/fsouza/go-dockerclient or  it would require to create an interface (may be with just Version function as you said) with docker client like what kube has with docker manager.

**Comment by aveshagarwal** [comment]
ok I am making some changes as what everyone suggested and push an update soon.

**Comment by aveshagarwal** [comment]
Pushed an update now.

File: pkg/cmd/server/kubernetes/node_config.go [hunk 4]
```diff
@@ -148,6 +151,12 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 	server.PodInfraContainerImage = imageTemplate.ExpandOrDie("pod")
 	server.CPUCFSQuota = true // enable cpu cfs quota enforcement by default
 
+	if dockerClient, _, err := docker.NewHelper().GetClient(); err == nil {
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/node_config.go]: @deads2k how do you feel about making server calls in BuildKubernetesNodeConfig? I'd prefer it not do that, but I also don't want config bits mutating after we build the config here

Comment (deads2k) [inline pkg/cmd/server/kubernetes/node_config.go]: > @deads2k how do you feel about making server calls in BuildKubernetesNodeConfig? I'd prefer it not do that, but I also don't want config bits mutating after we build the config here

I'm not a fan.  If this should be auto-detected, why don't we auto-detect it at the point of use?  If we don't want it auto-detected, why don't we have a config option?

Basically, this looks like its going to ignore any configuration option and automatically choose something appropriate, so why not move it closer to where its used.

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: @deads2k If I understood correctly, I dont think so it is ignoring any config option. The main thing it is doing is setting default for docker 1.9 to parallel image pulls. It keeps the behavior as it is for docker version <=1.8 and it also honors the user set value of serializeImagepulls irrespective of docker version, and thats why the changes are before the call to cmdflags.Resolve.

Comment (deads2k) [inline pkg/cmd/server/kubernetes/node_config.go]: > @deads2k If I understood correctly, I dont think so it is ignoring any config option. The main thing it is doing is setting default for docker 1.9 to parallel image pulls. It keeps the behavior as it is for docker version <=1.8 and it also honors the user set value of serializeImagepulls irrespective of docker version, and thats why the changes are before the call to cmdflags.Resolve.

Ok, I see that now.  I'm ok with doing this.  We transitively require a docker connection via `UnsecuredKubeleteConfig`, so this isn't much worse and its much better than the alternative.

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: FYI, we wanted to do this upstream as part of the Kubelet initialization process, but upstream rejected modifying the Kubelet to do this, and suggested instead that we modify the external bootstrapping code (salt). We are operating under the assumption that it will take significant bake-in time upstream to demonstrate that parallel pulls are safe again (we are optimistic/confident that the Docker 1.9 changes have fixed the concurrency issues), so we're making the change here in Origin instead.

**Comment by ncdc** [comment]
@liggitt @deads2k I noticed this file doesn't have any unit tests in the tree. Do you want one for this PR?

**Comment by liggitt** [comment]
I'm trying to envision how I'd want this tested

**Comment by ncdc** [comment]
It's probably not the easiest thing to do quickly. I'd want to make sure that based on whatever Docker API version we say exists in the test case, the default for serialized image pulls is set appropriately. But there's a lot of other stuff that goes on in this function...

File: pkg/cmd/server/kubernetes/node_config.go [hunk 5]
```diff
@@ -57,6 +58,8 @@ type NodeConfig struct {
 	FilteringEndpointsHandler osdnapi.FilteringEndpointsConfigHandler
 }
 
+const minimumDockerAPIVersionForParallelPulls = "1.21"
```
Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Please add a comment that this is Docker 1.9

File: pkg/cmd/util/docker/docker.go [hunk 6]
```diff
@@ -45,3 +45,28 @@ func (h *Helper) GetClientOrExit() (*docker.Client, string) {
 	}
 	return client, addr
 }
+
+// IsAPIVersionCompatible returns true if the current docker API version >= the input
+// version, otherwise it returns false.
+func IsAPIVersionCompatible(dc *docker.Client, version string) bool {
+	env, err := dc.Version()
+	if err != nil {
+		glog.Warningf("docker: failed to get docker version: %v", err)
```
Comment (ncdc) [inline pkg/cmd/util/docker/docker.go]: I don't think these log messages need to have the `docker: ` prefix in them

File: pkg/cmd/util/docker/docker.go [hunk 7]
```diff
@@ -45,3 +45,28 @@ func (h *Helper) GetClientOrExit() (*docker.Client, string) {
...
+		return false
+	}
+
+	apiVersion := env.Get("ApiVersion")
+	serverversion, err := docker.NewAPIVersion(apiVersion)
```
Comment (ncdc) [inline pkg/cmd/util/docker/docker.go]: serverVersion

**Comment by ncdc** [comment]
LGTM modulo the few comments I just added.

**Comment by aveshagarwal** [comment]
@ncdc thanks. Updated PR.

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
@smarterclayton PTAL

File: pkg/cmd/server/kubernetes/node_config.go [hunk 8]
```diff
@@ -148,6 +152,12 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 	server.PodInfraContainerImage = imageTemplate.ExpandOrDie("pod")
 	server.CPUCFSQuota = true // enable cpu cfs quota enforcement by default
 
+	if dockerClient, _, err := docker.NewHelper().GetClient(); err == nil {
+		if docker.IsAPIVersionCompatible(dockerClient, minimumDockerAPIVersionForParallelPulls) {
```
Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: I would prefer that this happened during EnsureDocker() and the server updated.

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: Is that possible?

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: See https://github.com/openshift/origin/pull/6920 for how I am refactoring Docker startup

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Would we need to add SerializeImagePulls to configapi.NodeConfig to be able to do that?

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: NodeConfig has KubeletServer

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: It has KubeletArguments, but we construct a new KubeletServer in this func (BuildKubernetesNodeConfig)

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: SerializeImagePulls is bool, and it wouldn't be possible to know if user changed it or not where Ensuredocker is called, that would mean mutating user configured value of SerializeImagePulls.

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Avesh brings up a good point - we may want to allow users to still serialize image pulls if they so choose.

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: I don't like calling docker before EnsureDocker - if ensure docker is
called from build* arguments that's acceptable.

On Feb 1, 2016, at 4:54 PM, Avesh Agarwal <notifications@github.com> wrote:

In pkg/cmd/server/kubernetes/node_config.go
<https://github.com/openshift/origin/pull/6538#discussion_r51487255>:

> @@ -148,6 +152,12 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
>  	server.PodInfraContainerImage = imageTemplate.ExpandOrDie("pod")
>  	server.CPUCFSQuota = true // enable cpu cfs quota enforcement by default
>
> +	if dockerClient, _, err := docker.NewHelper().GetClient(); err == nil {
> +		if docker.IsAPIVersionCompatible(dockerClient, minimumDockerAPIVersionForParallelPulls) {

SerializeImagePulls is bool, and it wouldn't be possible to know if user
changed it or not where Ensuredocker is called, that would mean mutating
user configured value of SerializeImagePulls.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6538/files#r51487255>.

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: Another option seems like to call EnsureDocker in BuildKubernetesNodeConfig. I think having a handle to docker client seems a reasonable part of building node config. Also currently, the calls to BuildKubernetesNodeConfig and EnsureDocker in StartNode() are not very far apart (there is only one EnsureVolumeDir() in between, so not much).

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: I agree with @smarterclayton. Calling ensuredocker in BuildKubernetesNodeConfig should be a better option.

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: Actually... Now that I think about it - that would break config
generation on a non-Docker system.  So we need a way to infer this
default in one path (startup) and skip it in another (config gen)

Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: Would it? This only sets serial to false if it's able to get the docker client and the version is ok. If it encounters an error, it keeps the default serial=true and moves on.

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: I guess.  I'm suddenly skeptical about defaulting based on detected docker
in something that is intended to be set by admins.  Why is this something
we are doing in our code, and not simply a config flag?

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: You mean a new config flag for setting docker version and then setting serializeimagepull based on that new flag? Somehow I think it could be more error prone even for admins, 1) what if the new flag and serializeimagepulls are configured incorrectly and not in accordance to each other, 2) what if docker is upgraded or downgraded and an admin forgets to reconfigure any of these flag.

I was thinking more about it and I am still wondering why is it an issue to determine docker version at run time (infact to avoid manual errors) and then setting config vars accordingly? I am still trying to understand if  there is any technical reason for that?

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: Is SerializeImagePulls *only* a workaround for Docker being broken in 1.8?
Is there ever any case an admin wants to force its use?

Comment (aveshagarwal) [inline pkg/cmd/server/kubernetes/node_config.go]: As I understand, SerializeImagePulls was introduced due to issue with 1.8 with parallel pulls. If parallel-image-pulls works fine with any/future version of docker, then why would someone want to use serialpulls, I am not sure, doesn't make much sense though.

Comment (smarterclayton) [inline pkg/cmd/server/kubernetes/node_config.go]: Ok, then i'm fine with this being very closely coupled here, and not end
user configurable.

**Comment by ncdc** [comment]
@smarterclayton @liggitt @deads2k is this ready to be merged, or is additional work needed?

File: pkg/cmd/util/docker/docker.go [hunk 9]
```diff
@@ -45,3 +45,28 @@ func (h *Helper) GetClientOrExit() (*docker.Client, string) {
...
+// version, otherwise it returns false.
+func IsAPIVersionCompatible(dc *docker.Client, version string) bool {
+	env, err := dc.Version()
+	if err != nil {
+		glog.Warningf("Failed to get docker server version: %v", err)
+		return false
+	}
+
+	apiVersion := env.Get("ApiVersion")
+	serverVersion, err := docker.NewAPIVersion(apiVersion)
```
Comment (liggitt) [inline pkg/cmd/util/docker/docker.go]: is this affected by any of the semver issues we saw?

File: pkg/cmd/util/docker/docker.go [hunk 10]
```diff
@@ -45,3 +45,28 @@ func (h *Helper) GetClientOrExit() (*docker.Client, string) {
...
+	if err != nil {
+		glog.Warningf("Failed to parse docker server version %q: %v", apiVersion, err)
+		return false
+	}
+
+	minimumVersion, err := docker.NewAPIVersion(version)
```
Comment (liggitt) [inline pkg/cmd/util/docker/docker.go]: same question here

File: pkg/cmd/util/docker/docker.go (hunk 9 above)
Comment (ncdc) [inline pkg/cmd/util/docker/docker.go]: Yes: `Unable to parse version "1.8.1.fc21": "fc21" is not an integer`

Upstream fix (internal to kubelet): https://github.com/kubernetes/kubernetes/pull/20827/. We'll probably need to do something similar.

Comment (ncdc) [inline pkg/cmd/util/docker/docker.go]: Oh wait, no. ApiVersion is not a semver. It's e.g. `1.21`.

**Comment by liggitt** [comment]
I'm still a little weirded out by making server calls when building config, unsure which calls have the potential to hang, but if @smarterclayton's ok with it, then I guess I am

**Comment by ncdc** [comment]
Alternatives:

1. Default to serial image pulls until Docker 1.8 is EOL; document what admins should do with Docker 1.9+
1. Assume most people will use 1.9+, so default to parallel; document what admins should do with 1.8
1. ???

**Comment by smarterclayton** [comment]
I'm ok with 2.

**Comment by ncdc** [comment]
We can't do 2 until RHEL has 1.9+.

**Comment by smarterclayton** [comment]
We can if the dates for 1.9 line up with the release the code is in.

**Comment by rhatdan** [comment]
Right now there is no plans to ship docker-1.9 for RHEL, only docker-1.10.

File: pkg/cmd/util/docker/docker.go (hunk 9 above)
Comment (aveshagarwal) [inline pkg/cmd/util/docker/docker.go]: Yes API version shouldn't be an issue.

**Comment by detiber** [comment]
>Right now there is no plans to ship docker-1.9 for RHEL, only docker-1.10.

This is no longer true, right? I was under the assumption that we were now going to ship 1.9

**Comment by aveshagarwal** [comment]
As we are simply disabling serial image pulls in origin by default irrespective of docker version: https://github.com/openshift/origin/pull/8363

I dont think this is needed any more, so closing this.
